### PR TITLE
Announcement Update (squashed)

### PR DESCRIPTION
**Pull request by peterpih**
Checks or overlapping dates in the Announcement by pairwise comparison of dates in each Announcement with the current Announcement being created

### CODE REVIEW

**Comment by peterpih** [comment]
Highlights:
helpers/announcement_helper.rb updated to 2 space indent and streamline refactoring (recovered from pervious commit)
models/announcement.rb refactored for pairwise checking of overlapping dates
models/announcement_spec.rb refactored to check for Announcement testing

**Comment by dinshaw** [comment]
Not squashed; still four commits in this PR. Should be one.

File: .gitignore [hunk 1]
```diff
@@ -21,6 +21,7 @@ config/database.yml
 config/mail.yml
 db/schema.rb
 db/*.sqlite3
+Gemfile.lock
```
Comment (dinshaw) [inline .gitignore]: This file cannot be ignored.

File: app/models/announcement.rb [hunk 2]
```diff
@@ -1,32 +1,31 @@
 class Announcement < ActiveRecord::Base
-  attr_accessible :start_date, :end_date, :button, :link, :body
-  
+  attr_accessible :id, :start_date, :end_date, :button, :link, :body
```
Comment (dinshaw) [inline app/models/announcement.rb]: id should not be in this list.

File: app/models/announcement.rb [hunk 3]
```diff
@@ -1,32 +1,31 @@
...
 
-  def date_range
-    start_date..end_date
-  end
-
   def overlapping_dates
-    announcements = Announcement.all
-    errors.add(:base, "Announcement conflicts with other announcements") if announcements.any? { |announcement| announcement.date_range.overlaps? date_range }
+    if start_date && end_date
+      Announcement.find do | d |
```
Comment (dinshaw) [inline app/models/announcement.rb]: What is the sequel this generates?

File: app/models/announcement.rb [hunk 4]
```diff
@@ -1,32 +1,31 @@
...
 
   def start_date_before_end_date
-    errors.add(:end_date, "cannot be earlier than start date") if start_date > end_date
+    if start_date && end_date && start_date > end_date
+      errors.add(:end_date, "cannot be earlier than start date") 
+    end
   end
-  
+
   def self.current
```
Comment (dinshaw) [inline app/models/announcement.rb]: Rails has some built in functionality for this sort of thing called `scopes`: http://guides.rubyonrails.org/active_record_querying.html#scopes

File: spec/factories/announcements.rb [hunk 5]
```diff
@@ -1,7 +1,11 @@
 FactoryGirl.define do
+  factory :announcement_empty do
+
+  end
   factory :announcement do
-    sequence(:start_date) { |n| Date.today + n.days }
-    sequence(:end_date) { |n| start_date + n.weeks }
+  	start_date { Date.today }
```
Comment (dinshaw) [inline spec/factories/announcements.rb]: indentation.

File: app/models/announcement.rb (hunk 3 above)
Comment (peterpih) [inline app/models/announcement.rb]: Looking at all the announcements:
`SELECT "announcements".* FROM "announcements"`

Comment (dinshaw) [inline app/models/announcement.rb]: No i mean the whole thing. Run a test that exercises that validation, and tail the test log.
If this is pure sql, then it is a good validation, however if it is instantiating every Announcement in the database as an ActiveModel class and calling their date attributes, then that is bad.

Comment (peterpih) [inline app/models/announcement.rb]: I've refactored to be more sql, one record gets instantiated in order to show the user a conflicting start_date and end_date

File: app/models/announcement.rb (hunk 2 above)
Comment (peterpih) [inline app/models/announcement.rb]: Removed.

File: .gitignore (hunk 1 above)
Comment (peterpih) [inline .gitignore]: Removed from .gitignore

File: spec/factories/announcements.rb (hunk 5 above)
Comment (peterpih) [inline spec/factories/announcements.rb]: Fixed.

**Comment by peterpih** [comment]
sql query for checking overlapping dates produces:
```
SELECT  "announcements".* FROM "announcements"  WHERE ((end_date >= '2015-10-07' AND start_date <= '2015-10-08' ) AND id is distinct from 27)  ORDER BY "announcements"."id" ASC LIMIT 1
```
It instantiates 1 record to be able to show the user which dates an existing announcement causes the overlap

**Comment by dinshaw** [comment]
Needs to be squashed again.

File: app/helpers/announcements_helper.rb [hunk 6]
```diff
@@ -4,10 +4,11 @@ def announcement_hidden?(announcement)
   end
 
   def current_announcement
+    return @current_announcement if defined? @current_announcement
     @current_announcement = Announcement.current
```
Comment (dinshaw) [inline app/helpers/announcements_helper.rb]: What is the point of this method?

File: doc/Annoucements.md [hunk 7]
```diff
@@ -0,0 +1,5 @@
+###Title
```
Comment (dinshaw) [inline doc/Annoucements.md]: Remove this file unless you are going to add value to it.

File: app/helpers/announcements_helper.rb (hunk 6 above)
Comment (peterpih) [inline app/helpers/announcements_helper.rb]: It is a memoization of the instance variable @current_annoucement, the reason `if defined?` is used is to get around `nil` which may happen if there isn't a **current* Announcement (one which has dates spanning `Date.today`

File: doc/Annoucements.md (hunk 7 above)
Comment (peterpih) [inline doc/Annoucements.md]: I've deleted this file

**Comment by peterpih** [comment]
Squashed to 1 commit

File: app/helpers/announcements_helper.rb (hunk 6 above)
Comment (dinshaw) [inline app/helpers/announcements_helper.rb]: Memoizing is what we need here, but this is not memoizing anything. Check your syntax against the docs.

File: app/helpers/announcements_helper.rb [hunk 8]
```diff
@@ -4,10 +4,11 @@ def announcement_hidden?(announcement)
...
   end
 
   def no_announcement
-  	current_announcement.nil? or announcement_hidden?(current_announcement)
+    current_announcement.none or announcement_hidden?(current_announcement)
```
Comment (dinshaw) [inline app/helpers/announcements_helper.rb]: I don't think `none` what you want here. All boolean methods in Rails terminate with a question mark.
Check the docs for what this is actually doing.http://api.rubyonrails.org/classes/ActiveRecord/QueryMethods.html

**Pull request merged**